[3.x] Safe `Camera::unproject_position()`

lawnjelly opened the pull request:
`unproject_position()` can fail in some circumstances, and this needs to be conveyed to calling code.

This seems to occur when the test position is on the focal plane of a perspective camera matrix. Under these conditions, the perspective divide is 0, and the unprojected position is undefined (potentially infinite from the viewport centre).

We provide a safe version of unproject that detects this condition and returns false, and in the main routine we attempt to recover and provide a sensible result, while printing a warning once.

Fixes #96028

## Notes
* Prior to 3.6, `unproject_position()` had a major bug whereby it would return NaNs when fed some input data.
* This was changed to be flagged, but user complained about the error message ("shoot the messenger" :grin: ).
* In this PR an alternative safe version is provided that can return failure.
* Calling code in the editor that caused the error to appear is changed to use the safe version.
* We also deal with test points that are close to the focal plane with an epsilon. This is to prevent near infinite return values and math exceptions in calling code in rare situations.

<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

lawnjelly committed:
Safe `Camera::unproject_position()`

`unproject_position()` can fail in some circumstances, and this needs to be conveyed to calling code.

Calinou commented on scene/3d/camera.cpp:
```diff
@@ -484,13 +484,20 @@ Point2 Camera::unproject_position(const Vector3 &p_pos) const {
+
+Point2 Camera::unproject_position(const Vector3 &p_pos) const {
+	Point2 res;
+	ERR_FAIL_COND_V_MSG(!safe_unproject_position(p_pos, res), res, "unproject_position() failed.");
```
Could the message describe usual reasons why it fails?

timothyqiu commented:
I'm not familiar with math. Why can't points located on the camera plane be unprojected in the first place? Is this an implementation limitation?

lawnjelly commented:
Your guess is as good as mine, I'm reverse engineering a solution, at the moment it seems an implementation limitation.

The entire (previous implementation) is:
```
Point2 Camera::unproject_position(const Vector3 &p_pos) const {
	ERR_FAIL_COND_V_MSG(!is_inside_tree(), Vector2(), "Camera is not inside scene.");

	Size2 viewport_size = get_viewport()->get_visible_rect().size;

	CameraMatrix cm;

	if (mode == PROJECTION_ORTHOGONAL) {
		cm.set_orthogonal(size, viewport_size.aspect(), near, far, keep_aspect == KEEP_WIDTH);
	} else {
		cm.set_perspective(fov, viewport_size.aspect(), near, far, keep_aspect == KEEP_WIDTH);
	}

	Plane p(get_camera_transform().xform_inv(p_pos), 1.0);

	p = cm.xform4(p);

	// Prevent divide by zero.
	// TODO : Investigate, this was causing Nans.
	ERR_FAIL_COND_V(p.d == 0, Point2());
	p.normal /= p.d;

	Point2 res;
	res.x = (p.normal.x * 0.5 + 0.5) * viewport_size.x;
	res.y = (-p.normal.y * 0.5 + 0.5) * viewport_size.y;

	return res;
}
```

To quote from the issue:

> I'm not super familiar with unprojection math, but it looks like
> ```
> 	Transform cm_tr = get_camera_transform();
> 	Plane p(cm_tr.xform_inv(p_pos), 1.0);
> ```
> The plane normal is undefined (0, 0, 0) because the point to unproject is (maybe) at the origin of the camera?

The plane is being defined by a normal and a distance. At some locations, the inverse transform of `p_pos` is 0,0,0, which is suspicious, as the plane is badly defined.

You would have to debug with the MRP in the issue, I honestly don't know the exact conditions causing it.

From the debugger at the point of failure (where d == 0):

```
p before cm applied = (0, 0, 0), 1
p = (0, 0, -0.1), 0
p_pos = 1.68, 1.91, 3.08
viewport_size = 1956,1254
```
For camera transform and cm you would have to debug as writing these could cause confusion.

lawnjelly commented:
Ok I finally made an effort and worked out how this routine is meant to work and what was going wrong. As I understand it, the error is as stated above, the test point is on the focal plane.

I've adjusted the PR slightly to attempt to return a sensible(ish) result, and changed the condition to a `WARN_PRINT_ONCE` to prevent error spam. Strictly speaking this is a calling error, but it likely that users won't understand why this can't be done, and will prefer some kind of result that won't be too unexpected from the calling code.

akien-mga reviewed (approved):
Seems ok to me, didn't review in depth.

lawnjelly commented:
Thanks!
